relay-kafka/src/config.rs

```diff
@@ -80,93 +80,62 @@ impl KafkaTopic {
-            KafkaTopic::Profiles => &self.profiles,
-            KafkaTopic::ReplayEvents => &self.replay_events,
-            KafkaTopic::ReplayRecordings => &self.replay_recordings,
-            KafkaTopic::Monitors => &self.monitors,
-            KafkaTopic::Spans => &self.spans,
-            KafkaTopic::MetricsSummaries => &self.metrics_summaries,
-            KafkaTopic::Cogs => &self.cogs,
+        /// Configuration for topics.
+        #[derive(Serialize, Deserialize, Debug)]
+        #[serde(default)]
+        pub struct TopicAssignments {
+            $(
+                /// Topic name for $default_topic
+                #[serde(alias = $default_topic)]
+                pub $field_name: TopicAssignment,
+            )*
         }
-    }
-}
 
-impl Default for TopicAssignments {
-    fn default() -> Self {
-        Self {
-            events: "ingest-events".to_owned().into(),
-            attachments: "ingest-attachments".to_owned().into(),
-            transactions: "ingest-transactions".to_owned().into(),
-            outcomes: "outcomes".to_owned().into(),
-            outcomes_billing: None,
-            metrics_sessions: "ingest-metrics".to_owned().into(),
-            metrics_generic: "ingest-performance-metrics".to_owned().into(),
-            profiles: "profiles".to_owned().into(),
-            replay_events: "ingest-replay-events".to_owned().into(),
-            replay_recordings: "ingest-replay-recordings".to_owned().into(),
-            monitors: "ingest-monitors".to_owned().into(),
-            spans: "snuba-spans".to_owned().into(),
-            metrics_summaries: "snuba-metrics-summaries".to_owned().into(),
-            cogs: "shared-resources-usage".to_owned().into(),
+        impl TopicAssignments{
+            /// Get a topic assignment by [`KafkaTopic`] value
+            #[must_use]
+            pub fn get(&self, kafka_topic: KafkaTopic) -> &TopicAssignment {
+                match kafka_topic {
+                    $(
+                        $kafka_topic => &self.$field_name,
+                    )*
+                }
+            }
+        }
+
+        impl Default for TopicAssignments {
+            fn default() -> Self {
+                Self {
+                    $(
+                        $field_name: $default_topic.to_owned().into(),
+                    )*
+
+                }
+            }
         }
+    };
+}
+
+define_topic_assignments! {
+    TopicAssignments {
+        events: (KafkaTopic::Events, "ingest-events"),
+        attachments: (KafkaTopic::Attachments, "ingest-attachments"),
+        transactions: (KafkaTopic::Transactions, "ingest-transactions"),
+        outcomes: (KafkaTopic::Outcomes, "outcomes"),
+        outcomes_billing: (KafkaTopic::OutcomesBilling, "outcomes-billing"),
+        metrics_sessions: (KafkaTopic::MetricsSessions, "ingest-metrics"),
+        metrics_generic: (KafkaTopic::MetricsGeneric, "ingest-performance-metrics"),
```

Comment:
I think we need to bring back the `metrics` and `metrics_transactions` aliases to ensure backward compatibility. Maybe we can tweak the macro to allow multiple aliases.

Reply:
What do we need the backward compatibility for? Do we think that self-hosted users would have used these aliases? And is it the intention to support that forever?

Reply:
I guess to not cause a breaking change if it is unnecessary? We can also call it out as a breaking change in the changelog, the number of users who configure topics manually is probably close to zero.

Reply:
I updated the changelog in this PR. Did you want one calling it out in self-hosted too?

My personal feelings on this are that making this more complicated in order to maintain compatibility with something that probably no one has used is not really worth it. I'm not sure how this is generally handled in Relay, but in sentry and snuba codebases at least, breaking compatibility is generally fine as long as we call it out (and we did already break compatibility when this same change was shipped there)